ui-tests/src/test/java/org/openmrs/reference/ManageProviderSchedulesTest.java

```diff
@@ -1,59 +1,64 @@
+ * Other possibilities are Rules or extending TestSetup. The latter possibilities might be applied
+ * in future if the test code become hard to maintain/read
  */
-public class ManageProviderSchedulesTest extends TestBase {
-    private HomePage homePage;
-    private HeaderPage headerPage;
-    private AppointmentBlocksPage appointmentBlocksPage;
-
-    @Before
-    public void setUp() throws Exception {
-        
-        homePage = new HomePage(page);
-        assertPage(homePage);
-        headerPage = new HeaderPage(driver);
-        appointmentBlocksPage = new AppointmentBlocksPage(page);
-    }
-
-    @Ignore //ignored due to RA-904
-    @Test
-    public void manageProviderSchedulesTest() throws Exception {
-        appointmentBlocksPage.goToAppointmentBlock();
-        appointmentBlocksPage.selectLocation("Isolation Ward");
-        appointmentBlocksPage.clickOnCurrentDay();
-        appointmentBlocksPage.selectLocationBlock("Isolation Ward");
-        appointmentBlocksPage.enterStartTime("08");
-        appointmentBlocksPage.enterService("gyne");
-        appointmentBlocksPage.enterProvider("Super User");
-        appointmentBlocksPage.clickOnSave();
-        assertNotNull("Gynecology", appointmentBlocksPage.CURRENT_DAY);
-        appointmentBlocksPage.findBlock();
-        appointmentBlocksPage.clickOnEdit();
-        appointmentBlocksPage.enterService("derma");
-        appointmentBlocksPage.clickOnSave();
-        assertNotNull("Dermatology", appointmentBlocksPage.CURRENT_DAY);
-        appointmentBlocksPage.findBlock();
-        appointmentBlocksPage.clickOnDelete();
-        appointmentBlocksPage.clickOnConfirmDelete();}
-
-    @After
-    public void tearDown() throws Exception {
-        headerPage.clickOnHomeIcon();
-        headerPage.logOut();
-    }
+public class ManageProviderSchedulesTest extends ReferenceApplicationTestBase {
+	
+	protected HomePage homePage;
+	
+	protected AppointmentBlocksPage appointmentBlocksPage;
+	
+	protected String locationName;
+	
+	protected String locationUuid;
+	
+	protected String provider;
+	
+	@Before
+	public void setUp() throws Exception {
+		homePage = new HomePage(page);
+		assertPage(homePage);
+		AppointmentSchedulingPage appointmentSchedulingPage = homePage.goToAppointmentScheduling();
+		appointmentBlocksPage = appointmentSchedulingPage.goToAppointmentBlock();
+		/*
+		 * TODO The following code should be re-nabled when bug AM-186 is fixed
+		 * Location might not be available when running this test, thus we create one
+		 * locationName = "Location-"+TestData.randomSuffix();
+		 * We must be sure that the location has been created, the test will fail otherwise
+		 * locationUuid = TestData.createLocation(locationName);
+		 * assertNotNull(locationUuid);
+		*/
+		locationName = "Isolation Ward";
+		provider = "Provider " + TestData.randomSuffix();
+	}
+	
+	@After
+	public void tearDown() throws Exception {
+		/*
+		 *  TODO delete the location. Re-enable when bug AM-186 is fixed.
```

Comment:
Dear @dkayiwa ,
hereafter my analysis on the usage of clickOnHomeIcon and logOut methods
a) Only 7 tests use them, 6 out of 7 are ignored. The only one currently executing and 
those methods is  RecordMergeIssueTest.  When I run locally this class I get an error due to clickOnHomeIcon,  see the trace below.
Hence we have in production a test which fails, it sounds weird to me, I don't think it is possible because I suppose you would have ignored it.
b)According to the [guidelines](https://wiki.openmrs.org/display/docs/Automated+Testing+Guidelines), Point 6, **You must not access Driver in a test** 
As you can notice in the tests that use those methods, they have to invoke the driver 
object. As far as I've understood it is discouraged (and I personally agree on that). 

Definitely, if I'm not wrong on a) and b) we should open a new issue to fix those methods
and maybe you can try to merge the tests in these pull request (which I eager to see if they work :)).
Conversely, I'll try to use those methods also in my classes albeit I have already tried them, unfortunately they fail for the same reason of RecordMergeIssueTest. 
The RecordMergeIssueTest execution trace follows. 
`java.lang.NullPointerException
	at java.lang.String.contains(String.java:2133)
	at org.openmrs.uitestframework.page.Page$1.apply(Page.java:48)
	at org.openmrs.uitestframework.page.Page$1.apply(Page.java:1)
	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:238)
	at org.openmrs.uitestframework.page.Page.waitForPage(Page.java:122)
	at org.openmrs.uitestframework.page.Page.findElement(Page.java:156)
	at org.openmrs.uitestframework.page.Page.clickOn(Page.java:197)
	at org.openmrs.reference.page.HeaderPage.clickOnHomeIcon(HeaderPage.java:17)
	at org.openmrs.reference.RecordMergeIssueTest.recordMergeIssueTest(RecordMergeIssueTest.java:58)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)

java.lang.NullPointerException
	at java.lang.String.contains(String.java:2133)
	at org.openmrs.uitestframework.page.Page$1.apply(Page.java:48)
	at org.openmrs.uitestframework.page.Page$1.apply(Page.java:1)
	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:238)
	at org.openmrs.uitestframework.page.Page.waitForPage(Page.java:122)
	at org.openmrs.uitestframework.page.Page.findElement(Page.java:156)
	at org.openmrs.uitestframework.page.Page.clickOn(Page.java:197)
	at org.openmrs.reference.page.HeaderPage.clickOnHomeIcon(HeaderPage.java:17)
	at org.openmrs.reference.RecordMergeIssueTest.tearDown(RecordMergeIssueTest.java:81)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:47)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:44)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)

org.openqa.selenium.remote.SessionNotFoundException: The FirefoxDriver cannot be used after quit() was called.
Build info: version: '2.52.0', revision: '4c2593cfc3689a7fcd7be52549167e5ccc93ad28', time: '2016-02-11 11:22:43'
System info: host: 'domenico-VirtualBox', ip: '127.0.1.1', os.name: 'Linux', os.arch: 'i386', os.version: '4.4.0-71-generic', java.version: '1.8.0_121'
Driver info: driver.version: FirefoxDriver
	at org.openqa.selenium.firefox.FirefoxDriver$LazyCommandExecutor.execute(FirefoxDriver.java:377)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:644)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:701)
	at org.openqa.selenium.firefox.FirefoxDriver.getScreenshotAs(FirefoxDriver.java:341)
	at org.openmrs.uitestframework.test.TestBase.takeScreenshot(TestBase.java:355)
	at org.openmrs.uitestframework.test.TestBase$1.failed(TestBase.java:99)
	at org.junit.rules.TestWatcher.failedQuietly(TestWatcher.java:84)
	at org.junit.rules.TestWatcher.access$300(TestWatcher.java:46)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:62)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:271)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:70)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:50)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.junit.runners.Suite.runChild(Suite.java:127)
	at org.junit.runners.Suite.runChild(Suite.java:26)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:238)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:63)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:236)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:53)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:229)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:309)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
`